scripts/hud/jump-timing.ts

```diff
@@ -0,0 +1,44 @@
+import { PanelHandler } from 'util/module-helpers';
+import { RegisterHUDPanelForGamemode } from 'util/register-for-gamemodes';
+import { GamemodeCategories, GamemodeCategory } from 'common/web';
+import { rgbaTupleLerp, tupleToRgbaString } from 'util/colors';
+
+@PanelHandler()
+class JumpTimingHandler {
+	readonly panels = {
+		container: $<Panel>('#JumpTimingContainer'),
+		earlyBar: $<ProgressBar>('#JumpEarlyBar'),
+		lateBar: $<ProgressBar>('#JumpLateBar'),
+		earlyLabel: $<Label>('#JumpEarlyLabel'),
+		lateLabel: $<Label>('#JumpLateLabel')
+	};
+
+	constructor() {
+		RegisterHUDPanelForGamemode({
+			gamemodes: GamemodeCategories.get(GamemodeCategory.BHOP), // TODO: Add stamina bhop game mode
+			handledEvents: [
+				{
+					event: 'JumpTimingDataUpdate',
+					panel: this.panels.container,
+					callback: (jumpTiming, maxTimingPenalty) => this.onJumpTimingUpdate(jumpTiming, maxTimingPenalty)
+				}
+			]
+		});
+	}
+
+	onJumpTimingUpdate(jumpTiming: number, maxTimingPenalty: number) {
+		// Blend between yellow and orange for early timing, and orange and red for late timing since you lose more speed on the ground from friction and stamina
+		const earlyTimingRatio = jumpTiming < 0 ? -jumpTiming / maxTimingPenalty : 0;
+		this.panels.earlyBar.value = earlyTimingRatio;
+		this.panels.earlyLabel.text = jumpTiming < 0 ? -jumpTiming : 0;
+		const colorLerp = (Math.abs(jumpTiming) - 1) / (maxTimingPenalty - 1);
+		const earlyTimingColor = rgbaTupleLerp([255, 255, 0, 255], [255, 165, 0, 255], colorLerp);
+		this.panels.earlyBar.GetChild(0).style.backgroundColor = tupleToRgbaString(earlyTimingColor);
```

Comment:
I would avoid doing this call into C++ every time when you could just store this on `panels` like everything else, can also give an information name like `earlyBarProgress` or something.